gem/impero_utils.py

```diff
@@ -194,15 +194,15 @@ def place_declarations(tree, temporaries, get_indices):
     :arg get_indices: callable mapping from GEM nodes to an ordering
                       of free indices
     """
-    temporaries_set = set(temporaries)
-    assert len(temporaries_set) == len(temporaries)
+    numbering = {t: n for n, t in enumerate(temporaries)}
```

Comment:
For deterministic code generation: create a `dict` that maps temporaries to their ordering...